packages/compile-solidity/src/compilerSupplier/index.ts

```diff
@@ -0,0 +1,117 @@
+      maxBuffer: number;
+    };
+  }>;
+
+  constructor({ events, solcConfig }) {
+    const { version, docker, compilerRoots, spawn } = solcConfig;
+    this.events = events;
+    this.version = version ? version : defaultSolcVersion;
+    this.docker = docker;
+    this.compilerRoots = compilerRoots;
+    this.strategyOptions = {};
+    if (version) this.strategyOptions.version = this.version;
+    if (docker) this.strategyOptions.docker = compilerRoots;
+    if (compilerRoots) this.strategyOptions.compilerRoots = compilerRoots;
+    if (events) this.strategyOptions.events = events;
+    if (spawn) this.strategyOptions.spawn = spawn;
+  }
+
+  async load() {
+    const userSpecification = this.version;
+
+    let strategy;
+    const useDocker = this.docker;
+    const useNative = userSpecification === "native";
+    const useSpecifiedLocal =
+      userSpecification && this.fileExists(userSpecification);
+    const isValidVersionRange = semver.validRange(userSpecification);
+
+    if (useDocker) {
+      strategy = new Docker(this.strategyOptions);
+    } else if (useNative) {
+      strategy = new Native();
+    } else if (useSpecifiedLocal) {
+      strategy = new Local();
+    } else if (isValidVersionRange) {
+      strategy = new VersionRange(this.strategyOptions);
+    }
+
+    if (strategy) {
+      const solc = await strategy.load(userSpecification);
+      return { solc };
+    } else {
+      throw this.badInputError(userSpecification);
+    }
+  }
+
+  async list() {
+    const userSpecification = this.version;
+
+    let strategy;
+    const useDocker = this.docker;
+    const useNative = userSpecification === "native";
+    const useSpecifiedLocal =
+      userSpecification && this.fileExists(userSpecification);
+    const isValidVersionRange = semver.validRange(userSpecification);
+
+    if (useDocker) {
+      strategy = new Docker(this.strategyOptions);
+    } else if (useNative) {
+      strategy = new Native();
+    } else if (useSpecifiedLocal) {
+      strategy = new Local();
+    } else if (isValidVersionRange) {
+      strategy = new VersionRange(this.strategyOptions);
+    }
+
+    if (!strategy) {
+      throw this.badInputError(userSpecification);
+    }
+
+    if (!strategy.list) {
+      throw new Error(
+        `Cannot list versions for strategy ${strategy.constructor.name}`
+      );
+    }
+
+    return await strategy.list();
+  }
+
+  static getDefaultVersion() {
+    return defaultSolcVersion;
+  }
+
+  badInputError(userSpecification) {
+    const message =
+      `Could not find a compiler version matching ${userSpecification}. ` +
+      `compilers.solc.version option must be a string specifying:\n` +
+      `   - a path to a locally installed solcjs\n` +
+      `   - a solc version or range (ex: '0.4.22' or '^0.5.0')\n` +
+      `   - a docker image name (ex: 'stable')\n` +
+      `   - 'native' to use natively installed solc\n`;
+    return new Error(message);
+  }
+
+  fileExists(localPath) {
```

Comment:
@haltman-at notes [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687277390) that this method name is terrible.

Reply:
(I've inlined this)